src/Storages/PartitionStrategy.h

```diff
@@ -0,0 +1,91 @@
+#pragma once
+
+#include <Interpreters/ExpressionActions.h>
+#include <Parsers/IAST_fwd.h>
+
+#include <Processors/Chunk.h>
+
+namespace DB
+{
+
+struct PartitionStrategy
+{
+    struct PartitionExpressionActionsAndColumnName
+    {
+        ExpressionActionsPtr actions;
+        std::string column_name;
+    };
+
+    PartitionStrategy(ASTPtr partition_by_, const Block & sample_block_, ContextPtr context_);
+
+    virtual ~PartitionStrategy() = default;
+
+    virtual PartitionExpressionActionsAndColumnName getExpression() = 0;
+    virtual std::string getPath(const std::string & prefix, const std::string & partition_key) = 0;
+
+    virtual Chunk getChunkWithoutPartitionColumnsIfNeeded(const Chunk & chunk)
```

Comment:
This looks very shitty imo

Reply:
How is it different from `Chunk::clone` or `Chunk::cloneEmptyColumns` ?